### PR TITLE
[IMP] google_calendar: add single event synchronization

### PR DESCRIPTION
**Pull request by amdi-odoo**
Purpose
=======
Add the possibility to only synchronize one specific event of a user calendar instead of syncing the whole calendar.

See sub commits for more details.

Task-3285601

### CODE REVIEW

**Review by tde-banana-odoo — APPROVED** [review]

**Review by arj-odoo — CHANGES_REQUESTED** [review]
Hello :wave: 
Thanks for your PR. I understand the [need of this](https://github.com/odoo/enterprise/pull/46688/files#diff-2a0b9b70f45b12384f2fb4261b14fb06d4570cae1e7f10bba00096dff36d53b8R17) but I have a few questions.

For now, the event_id parameter of _sync_google_calendar is only filled in your appointment PR but it could change in the future. As you can see, _sync_google_calendar is now used to sync potentially a lot of events and recurrences.

I fear that in the future, we could use it to sync a single recurrence and it could cause issues. So I would make your code verify that event_id is is a single event calender.event , (not a calendar.recurrence) and that this event it does not belong to a recurrence.
google_id can be used to identify a calendar or a recurrence. We have helpers to make sure your event is single. All your appointment events are single so it won't break your specs. 

From what I understand, calling it from your write will get the single event values from google and sync it into Odoo but _sync_google_calendar does more than that. It will also
sync from Odoo to Google. If a lot of events needs to be push, your write in appointment will be quite slow. Please also note that these sync are normally performed
- once the calendar view is displayed
- in a post commit hook. As you call the api before super in appointment, if super raise an error, it will be rollback, your changes will be lost but google may have receive the request.

To avoid that, I would create a small method that would synchronize a single event.
It would:
- not work in full sync mode. Because you don't want to sync everything in your usecase. except InvalidSyncToken: would ask to refresh the user token (i.e. go the the calendar page and display the events)
- get the google values of that single event and write the values locally, then push the changes to google.

```py
try:
    event, next_sync_token, default_reminders =  calendar_service.get_events(self.google_calendar_account_id.calendar_sync_token, token=token, event_id=event_id)
except InvalidSyncToken:
    raise("Boum, make sure to do a full sync before")
# Google -> Odoo
send_updates = False
event.clear_type_ambiguity(self.env)
synced_events = self.env['calendar.event']._sync_google2odoo(event, default_reminders=())
# Odoo -> Google
odoo_event.with_context(send_updates=send_updates)._sync_odoo2google(calendar_service)
            
```
I did not tested this code but the idea is to do the minimal amount of things to sync that single event. What do you think ?

**Comment by amdi-odoo** [comment]
Hello @arj-odoo :smile: 
Thanks for the thorough explanations :smile: !! It really helps!
Indeed doing the minimal amount of things is the way to go :sweat_smile: 
I'll look into it and try to improve this part :slightly_smiling_face: :+1:

**Comment by arj-odoo** [comment]
Now that I think about it, the full_sync scenario is not an issue with a small method. You just have to make sure to update `calendar_sync_token = next_sync_token` when you get `next_sync_token` from the api.

**Comment by amdi-odoo** [comment]
Hi again :slightly_smiling_face: 

When calling this url for multiple events : `/calendar/v3/calendars/primary/events`
The API returns a list of GoogleEvent, the default_reminders and the nextSyncToken
cf https://developers.google.com/calendar/api/v3/reference/events/list

But when calling this link for a single event: `/calendar/v3/calendars/primary/events/**event_id**` 
We only get a single GoogleEvent from the API (no default_reminders and no nextSyncToken)
cf : https://developers.google.com/calendar/api/v3/reference/events/get
That's why I set those 2 parameters to their default values before returning:
```python
if event_id:
    next_sync_token = None
    default_reminders = ()
```
I don't think we need a new sync token in this case ? What do you think ? :thinking:

**Review by arj-odoo — APPROVED** [review]
Thanks for your work. I am ok with the changes here.

File: addons/google_calendar/models/calendar.py [hunk 1]
```diff
@@ -32,6 +35,13 @@ def _compute_google_id(self):
             elif not event.google_id:
                 event.google_id = False
 
+    @api.depends('videocall_location')
+    def _compute_videocall_source(self):
+        events_with_google_url = self.filtered(lambda event:
+            event.videocall_location and self.MEET_ROUTE in event.videocall_location)
```
Comment (tde-banana-odoo) [inline addons/google_calendar/models/calendar.py:41]: Maybe could be simplified a bit like
```suggestion
        events_with_google_url = self.filtered(lambda event: self.MEET_ROUTE in (event.videocall_location or ''))
```

File: addons/google_calendar/models/res_users.py [hunk 2]
```diff
@@ -53,24 +53,10 @@ def _get_google_sync_status(self):
...
-        full_sync = not bool(self.google_calendar_sync_token)
-        with google_calendar_token(self) as token:
-            try:
-                events, next_sync_token, default_reminders = calendar_service.get_events(self.google_calendar_account_id.calendar_sync_token, token=token)
-            except InvalidSyncToken:
-                events, next_sync_token, default_reminders = calendar_service.get_events(token=token)
-                full_sync = True
-        self.google_calendar_account_id.calendar_sync_token = next_sync_token
-
+        events, default_reminders, full_sync = results.values()
```
Comment (tde-banana-odoo) [inline addons/google_calendar/models/res_users.py:59]: Did not know this would work ^^

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, few comments but otherwise LGTM, thanks for your work and @arj-odoo 's review :)

File: addons/google_calendar/models/calendar.py [hunk 3]
```diff
@@ -14,10 +14,13 @@ class Meeting(models.Model):
     _name = 'calendar.event'
     _inherit = ['calendar.event', 'google.calendar.sync']
 
+    MEET_ROUTE = 'meet.google.com'
+
     google_id = fields.Char(
         'Google Calendar Event Id', compute='_compute_google_id', store=True, readonly=False)
     guests_readonly = fields.Boolean(
         'Guests Event Modification Permission', default=False)
+    videocall_source = fields.Selection(selection_add=[('google_meet', 'Google Meet')], ondelete={'google_meet': 'set discuss'})
```
[the remark this answers is not in the document]
Comment (amdi-odoo) [reply addons/google_calendar/models/calendar.py:23]: I think we need to keep set discuss as this is the `google_calendar` module so it is inheriting from the model of the `calendar` module and there is no default value for the `videocall_source` field in calendar. :thinking: 
```python
videocall_source = fields.Selection([('discuss', 'Discuss'), ('custom', 'Custom')], compute='_compute_videocall_source')
```
The default value is set to discuss only in `appointment`.

[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/google_calendar/models/calendar.py:23]: Ah ok then :)

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge

And hop !

**Pull request closed without merging**